Market doesn't update with termination set to a vega timestamp in the past

# Problem encountered
When updating data source filter to a Vega time value in the past, the market should terminate immediately.

# Observed behaviour
A market update is not possible when updating `oracle_termination_value` to a value in the past

## Automation
https://github.com/vegaprotocol/system-tests/pull/1104/files#diff-3a57b618abab501b8ddef8340638bcc4c06961c30d8346f66d83ded231c6704dR179

## Logs
```
context = {'AssetID': 'c07798bc90cf68ad2b5e4939b9617dd4d9f4dbc8eaa9ecd883675bd2f99a3a88', 'app_state.network_parameters': {'bloc...ed09c67fc917d7d43acc7230ebee7988ef06a0c2e31a569b328a7ab22a30e35f': '0xEe7D375bcB50C26d52E1A4a472D8822A2A22d94F'}], ...}, dsri_proposer = 'A4586F3587'
liq_prov = {'bshape': [[2, 'bestBidPrice', 1]], 'commitment': 3905000000000000000000000, 'fee': '0.3', 'shape': [[13, 'bestOfferPrice', 1]]}, dsri_traders = ['673A63EE20', '1956BCA985']

    @pytest.mark.last
    @pytest.mark.full
    @pytest.mark.oracles
    def test_update_term_past_trigger(context,
                                      dsri_proposer,
                                      liq_prov,
                                      dsri_traders):
        """
        Change a cash settled futures market so the trading terminated trigger source becomes a time triggered blank
        Value data source (or one with any value, to be discarded) with the trigger time being in the past.
        The market state changes to trading terminated immediately.
        0048-DSRI-012
        """
    
        proposer = dsri_proposer
        # create internal oracle with vega timestamp for termination in the future
        trading_stamp = str(int(time.time()) + 500)
        oracle = governMod.OracleInternal(context=context,
                                          public_key=proposer,
                                          termination_condition='GtThanEqual',
                                          termination_value=trading_stamp
                                          )
    
        _market = governMod.Market(context, proposer=proposer, oracle=oracle)
        _market.propose(proposer)
    
        # Take market into continuous trading
        ordersMod.reset_market(context, _market.id, keep_orders=True)
    
        # update market termination trigger to a past time
        updated_trading_stamp = str(int(time.time()) - 10)
        _update = governMod.UpdateMarket(context, marketId=_market.id, proposer=proposer,
                                         oracle_termination_value=updated_trading_stamp,
                                         oracle_termination_condition='GtThanEqual')
>       _update.propose(proposer)

tests/data_sources/dsri_test.py:210: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
modules/govMod/governMod.py:154: in propose
    self.check_event_exception(first_proposal_event, proposal_exception_expected,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <modules.govMod.governMod.UpdateMarket object at 0x7fb0c1174fd0>, events_found = [], exception_expected = False, event_type = 'Proposal'

    def check_event_exception(self, events_found, exception_expected, event_type=None):
        """
        handle proposal/vote event exceptions / dump events / assert or return
    
        exception_expected
        True - we expect the proposal workflow to fail so dump logs and return tx error to calling test if found (else not found we fail)
        False - we do not expect proposal workflow to fail, dump logs and raise error
        None - there may be an exception and it should be returned to calling test if found, do not raise error
        """
        assert event_type in ['Vote', 'Proposal']
        if len(events_found) == 0:
            logging.info(f"Event missing - dumping all events found for ref {self.ref} and checking for failures")
            all_events = utils.scan_events(busEventTypeDict.get(event_type)["Name"],
                                           search_list=[],
                                           ref=self.ref)
            [logging.info(f"Event Found--->{e}") for e in all_events]
    
            # check for tx errors using last wallet hash
            txerr_event = utils.find_tx_error_event_by_hash(self.wallet_response['txHash'])
            logging.info(f"tx_error--->{txerr_event}")
            if len(txerr_event) > 0:  # we dont always get a tx_error event so allow for this case
                self.tx_error = txerr_event
                self.tx_error_msg = json.loads(self.tx_error)['txErrEvent']['errMsg']
            else:
                self.tx_error = "No events and no tx_error event found- possible spam pre block detected"
            for event in all_events:
                found_event = json.loads(event)
                state = found_event['proposal']['state']
                if state == 'STATE_REJECTED':
                    self.rejection_reason = found_event['proposal']['reason']
                    self.error_details = found_event['proposal']['errorDetails']
                elif state == 'STATE_DECLINED':
                    self.error_details = found_event['proposal']['errorDetails']
                logging.info(self.tx_error)
            if exception_expected is False:
>               raise AssertionError(
                    f"Exception is not expected but occurred anyway...\n\n{self.rejection_reason} {self.error_details}")
E               AssertionError: Exception is not expected but occurred anyway...
E               
E               PROPOSAL_ERROR_INVALID_FUTURE_PRODUCT oracle spec termination time before enactment

modules/govMod/governMod.py:382: AssertionError

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -240,6 +240,7 @@ The init command now also generate the configuration for tendermint, the flags `
 - [5304](https://github.com/vegaprotocol/vega/issues/5304) - Fix bug causing trade events at auction end showing the wrong price.
 - [5345](https://github.com/vegaprotocol/vega/issues/5345) - Fix issue with state variable transactions assumed gone missing
 - [5351](https://github.com/vegaprotocol/vega/issues/5351) - Fix panic when node is interrupted before snapshot engine gets cleared and initialised
+- [5972](https://github.com/vegaprotocol/vega/issues/5972) - Allow submitting a market update with termination oracle ticking before enactment of the update
 
 ## 0.50.2
 

--- a/core/execution/market_test.go
+++ b/core/execution/market_test.go
@@ -123,6 +123,7 @@ type testMarket struct {
 	mktCfg           *types.Market
 	oracleEngine     *oracles.Engine
 	stateVar         *stubs.StateVarStub
+	builtinOracle    *oracles.Builtin
 
 	// Options
 	Assets []types.Asset
@@ -164,7 +165,7 @@ func newTestMarket(t *testing.T, now time.Time) *testMarket {
 
 	tm.broker.EXPECT().Send(gomock.Any()).AnyTimes().Do(eventFn)
 	tm.broker.EXPECT().SendBatch(gomock.Any()).AnyTimes().Do(eventsFn)
-
+	tm.builtinOracle = oracles.NewBuiltinOracle(tm.oracleEngine, tm.timeService)
 	return tm
 }
 
@@ -369,6 +370,7 @@ func getTestMarket2WithDP(
 
 	oracleEngine := oracles.NewEngine(log, oracles.NewDefaultConfig(), timeService, broker)
 	tm.oracleEngine = oracleEngine
+	tm.builtinOracle = oracles.NewBuiltinOracle(tm.oracleEngine, tm.timeService)
 
 	// add the token asset
 	tokAsset := types.Asset{
@@ -985,6 +987,116 @@ func TestMarketWithTradeClosing(t *testing.T) {
 
 	tm.now = futureTime
 	closed := tm.market.OnTick(ctx, futureTime)
+	assert.True(t, closed)
+}
+
+func TestUpdateMarketWithOracleSpecEarlyTermination(t *testing.T) {
+	party1 := "party1"
+	party2 := "party2"
+	now := time.Unix(10, 0)
+	closingAt := time.Unix(20, 0)
+	tm := getTestMarket(t, now, nil, nil)
+	ctx := vegacontext.WithTraceID(context.Background(), vgcrypto.RandomHash())
+	defer tm.ctrl.Finish()
+	// add 2 parties to the party engine
+	// this will create 2 parties, credit their account
+	// and move some monies to the market
+	// this will also output the closed accounts
+	addAccount(t, tm, party1)
+	addAccount(t, tm, party2)
+
+	// submit orders
+	// party1 buys
+	// party2 sells
+	orderBuy := &types.Order{
+		Type:        types.OrderTypeLimit,
+		TimeInForce: types.OrderTimeInForceGTT,
+		Status:      types.OrderStatusActive,
+		ID:          "",
+		Side:        types.SideBuy,
+		Party:       party1,
+		MarketID:    tm.market.GetID(),
+		Size:        100,
+		Price:       num.NewUint(100),
+		Remaining:   100,
+		CreatedAt:   now.UnixNano(),
+		ExpiresAt:   closingAt.UnixNano(),
+		Reference:   "party1-buy-order",
+	}
+	orderSell := &types.Order{
+		Type:        types.OrderTypeLimit,
+		TimeInForce: types.OrderTimeInForceGTT,
+		Status:      types.OrderStatusActive,
+		ID:          "",
+		Side:        types.SideSell,
+		Party:       party2,
+		MarketID:    tm.market.GetID(),
+		Size:        100,
+		Price:       num.NewUint(100),
+		Remaining:   100,
+		CreatedAt:   now.UnixNano(),
+		ExpiresAt:   closingAt.UnixNano(),
+		Reference:   "party2-sell-order",
+	}
+
+	// submit orders
+	tm.broker.EXPECT().Send(gomock.Any()).AnyTimes()
+	// tm.transferResponseStore.EXPECT().Add(gomock.Any()).AnyTimes()
+
+	fmt.Printf("%s\n", orderBuy.String())
+	_, err := tm.market.SubmitOrder(ctx, orderBuy)
+	assert.Nil(t, err)
+	if err != nil {
+		t.Fail()
+	}
+	tm.now = tm.now.Add(time.Second)
+	tm.market.OnTick(ctx, tm.now)
+	fmt.Printf("%s\n", orderBuy.String())
+	require.Equal(t, types.MarketStateSuspended, tm.market.State()) // enter auction
+
+	_, err = tm.market.SubmitOrder(ctx, orderSell)
+	assert.Nil(t, err)
+	if err != nil {
+		t.Fail()
+	}
+	tm.now = tm.now.Add(time.Second)
+	tm.market.OnTick(ctx, tm.now)
+
+	// now update the market
+	updatedMkt := tm.mktCfg.DeepClone()
+
+	updatedMkt.TradableInstrument.Instrument.GetFuture().OracleSpecForTradingTermination.Filters = []*types.OracleSpecFilter{
+		{
+			Key: &types.OracleSpecPropertyKey{
+				Name: oracles.BuiltinOracleTimestamp,
+				Type: oraclespb.PropertyKey_TYPE_TIMESTAMP,
+			},
+			Conditions: []*types.OracleSpecCondition{
+				{
+					Operator: oraclespb.Condition_OPERATOR_GREATER_THAN_OR_EQUAL,
+					Value:    "0",
+				},
+			},
+		},
+	}
+	updatedMkt.TradableInstrument.Instrument.GetFuture().OracleSpecBinding.TradingTerminationProperty = oracles.BuiltinOracleTimestamp
+
+	err = tm.market.Update(context.Background(), updatedMkt, tm.oracleEngine)
+	require.NoError(t, err)
+	tm.builtinOracle.OnTick(ctx, tm.now)
+	tm.market.OnTick(ctx, tm.now)
+	require.Equal(t, types.MarketStateTradingTerminated, tm.market.State())
+
+	properties := map[string]string{}
+	properties["prices.ETH.value"] = "100"
+	err = tm.oracleEngine.BroadcastData(ctx, oracles.OracleData{
+		PubKeys: []string{"0xDEADBEEF"},
+		Data:    properties,
+	})
+	require.NoError(t, err)
+
+	tm.now = tm.now.Add(time.Second)
+	closed := tm.market.OnTick(ctx, tm.now)
 	assert.True(t, closed)
 }
 

--- a/core/governance/engine.go
+++ b/core/governance/engine.go
@@ -843,6 +843,7 @@ func (e *Engine) validateChange(terms *types.ProposalTerms) (types.ProposalError
 		closeTime := time.Unix(terms.ClosingTimestamp, 0)
 		return validateNewMarketChange(terms.GetNewMarket(), e.assets, true, e.netp, enactTime.Sub(closeTime), enct)
 	case types.ProposalTermsTypeUpdateMarket:
+		enct.shouldNotVerify = true
 		return validateUpdateMarketChange(terms.GetUpdateMarket(), enct)
 	case types.ProposalTermsTypeNewAsset:
 		return terms.GetNewAsset().Validate()

--- a/core/governance/engine_test.go
+++ b/core/governance/engine_test.go
@@ -62,7 +62,6 @@ func TestSubmitProposals(t *testing.T) {
 	t.Run("Submitting a proposal without enough stake fails", testSubmittingProposalWithoutEnoughStakeFails)
 
 	t.Run("Submitting a time-triggered proposal for new market with termination time before enactment time fails", testSumittingTimeTriggeredProposalNewMarketTerminationBeforeEnactmentFails)
-	t.Run("Submitting a time-triggered proposal for market update with termination time same as enactment time fails", testSubmittingTimeTriggeredProposalMarketUpdateTerminationEnactmentFails)
 
 	t.Run("Voting on non-existing proposal fails", testVotingOnNonExistingProposalFails)
 	t.Run("Voting with non-existing account fails", testVotingWithNonExistingAccountFails)
@@ -361,91 +360,6 @@ func testSumittingTimeTriggeredProposalNewMarketTerminationBeforeEnactmentFails(
 	eng.expectRejectedProposalEvent(t, proposer, proposal2.ID, types.ProposalErrorInvalidFutureProduct)
 
 	_, err = eng.submitProposal(t, proposal2)
-	require.Error(t, err)
-	assert.Contains(t, err.Error(), "oracle spec termination time before enactment")
-}
-
-func testSubmittingTimeTriggeredProposalMarketUpdateTerminationEnactmentFails(t *testing.T) {
-	eng := getTestEngine(t)
-	defer eng.ctrl.Finish()
-
-	proposer := vgrand.RandomStr(5)
-	now := eng.tsvc.GetTimeNow()
-
-	// Create new market with termination time 1 hour after enactment time
-	// Enactment time for new market is now + 96 hours
-	filter, binding := produceTimeTriggeredOracleSpec(now.Add(2 * 48 * time.Hour).Add(1 * time.Minute))
-	proposal1 := eng.newProposalForNewMarket(proposer, now, filter, binding)
-
-	eng.ensureTokenBalanceForParty(t, proposer, 1)
-	eng.ensureStakingAssetTotalSupply(t, 9)
-	eng.ensureAllAssetEnabled(t)
-
-	eng.expectOpenProposalEvent(t, proposer, proposal1.ID)
-
-	_, err := eng.submitProposal(t, proposal1)
-	require.NoError(t, err)
-
-	voter1 := vgrand.RandomStr(5)
-	eng.ensureTokenBalanceForParty(t, voter1, 7)
-	eng.expectVoteEvent(t, voter1, proposal1.ID)
-
-	err = eng.addYesVote(t, voter1, proposal1.ID)
-	require.NoError(t, err)
-
-	eng.ensureTokenBalanceForParty(t, voter1, 7)
-	eng.expectTotalGovernanceTokenFromVoteEvents(t, "1", "7")
-	eng.expectPassedProposalEvent(t, proposal1.ID)
-	eng.markets.EXPECT().GetMarketState(proposal1.ID).Return(types.MarketStateClosed, nil)
-
-	// Close the voting for proposal 1
-	afterClosing := time.Unix(proposal1.Terms.ClosingTimestamp, 0).Add(time.Second)
-	toBeEnacted, toBeClosed := eng.OnTick(context.Background(), afterClosing)
-	require.Len(t, toBeClosed, 1)
-	require.Len(t, toBeEnacted, 0)
-
-	afterEnactment := time.Unix(proposal1.Terms.EnactmentTimestamp, 0).Add(time.Second)
-	toBeEnacted, toBeClosed = eng.OnTick(context.Background(), afterEnactment)
-	require.Len(t, toBeEnacted, 1)
-	require.Len(t, toBeClosed, 0)
-	assert.Equal(t, proposal1.ID, toBeEnacted[0].Proposal().ID)
-
-	voter2 := vgrand.RandomStr(5)
-	err = eng.addNoVote(t, voter2, proposal1.ID)
-	require.Error(t, err)
-	assert.EqualError(t, err, governance.ErrProposalDoesNotExist.Error())
-
-	// Voting for proposal 1 is closed, we are able to place another proposal now
-	// Set termination timestamp 45 minutes before enactment time
-	// Enactment time for updating market is now + 192 hours
-	filter, binding = produceTimeTriggeredOracleSpec(now.Add(3 * 48 * time.Hour).Add(1 * 47 * time.Hour).Add(15 * time.Minute))
-	proposal2 := eng.newProposalForMarketUpdate("market-1", proposer, now, filter, binding)
-
-	eng.ensureTokenBalanceForParty(t, proposer, 1)
-	eng.ensureAllAssetEnabled(t)
-
-	marketID := proposal2.MarketUpdate().MarketID
-	eng.ensureExistingMarket(t, marketID)
-	eng.ensureEquityLikeShareForMarketAndParty(t, marketID, proposer, 0.1)
-
-	eng.expectRejectedProposalEvent(t, proposer, proposal2.ID, types.ProposalErrorInvalidFutureProduct)
-	_, err = eng.submitProposal(t, proposal2)
-	require.Error(t, err)
-	assert.Contains(t, err.Error(), "oracle spec termination time before enactment")
-
-	// Set termination timestamp same as enactment timestamp
-	filter, binding = produceTimeTriggeredOracleSpec(now.Add(4 * 48 * time.Hour))
-	proposal3 := eng.newProposalForMarketUpdate("market-1", proposer, now, filter, binding)
-
-	eng.ensureTokenBalanceForParty(t, proposer, 1)
-	eng.ensureAllAssetEnabled(t)
-
-	marketID = proposal3.MarketUpdate().MarketID
-	eng.ensureExistingMarket(t, marketID)
-	eng.ensureEquityLikeShareForMarketAndParty(t, marketID, proposer, 0.1)
-
-	eng.expectRejectedProposalEvent(t, proposer, proposal3.ID, types.ProposalErrorInvalidFutureProduct)
-	_, err = eng.submitProposal(t, proposal3)
 	require.Error(t, err)
 	assert.Contains(t, err.Error(), "oracle spec termination time before enactment")
 }

--- a/core/governance/engine_update_market_test.go
+++ b/core/governance/engine_update_market_test.go
@@ -21,6 +21,7 @@ import (
 	"code.vegaprotocol.io/vega/core/netparams"
 	"code.vegaprotocol.io/vega/core/types"
 	vgrand "code.vegaprotocol.io/vega/libs/rand"
+	oraclespb "code.vegaprotocol.io/vega/protos/vega/oracles/v1"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -49,6 +50,45 @@ func testSubmittingProposalForMarketUpdateSucceeds(t *testing.T) {
 	proposer := vgrand.RandomStr(5)
 	proposal := eng.newProposalForMarketUpdate("market-1", proposer, eng.tsvc.GetTimeNow(), nil, nil)
 	marketID := proposal.MarketUpdate().MarketID
+
+	// setup
+	eng.ensureTokenBalanceForParty(t, proposer, 1000)
+	eng.ensureEquityLikeShareForMarketAndParty(t, marketID, proposer, 0.1)
+	eng.ensureExistingMarket(t, marketID)
+
+	// expect
+	eng.expectOpenProposalEvent(t, proposer, proposal.ID)
+
+	// when
+	toSubmit, err := eng.submitProposal(t, proposal)
+
+	// then
+	require.NoError(t, err)
+	require.NotNil(t, toSubmit)
+}
+
+func TestSubmittingProposalForMarketUpdateWithEarlyTerminationSucceeds(t *testing.T) {
+	eng := getTestEngine(t)
+	defer eng.ctrl.Finish()
+
+	// given
+	proposer := vgrand.RandomStr(5)
+	proposal := eng.newProposalForMarketUpdate("market-1", proposer, eng.tsvc.GetTimeNow(), nil, nil)
+	marketID := proposal.MarketUpdate().MarketID
+
+	proposal.Terms.Change.(*types.ProposalTermsUpdateMarket).UpdateMarket.Changes.Instrument.Product.(*types.UpdateInstrumentConfigurationFuture).Future.OracleSpecForTradingTermination.Filters = []*types.OracleSpecFilter{{
+		Key: &types.OracleSpecPropertyKey{
+			Name: "vegaprotocol.builtin.timestamp",
+			Type: oraclespb.PropertyKey_TYPE_TIMESTAMP,
+		},
+		Conditions: []*types.OracleSpecCondition{
+			{
+				Operator: oraclespb.Condition_OPERATOR_GREATER_THAN_OR_EQUAL,
+				Value:    "0",
+			},
+		},
+	}}
+	proposal.Terms.Change.(*types.ProposalTermsUpdateMarket).UpdateMarket.Changes.Instrument.Product.(*types.UpdateInstrumentConfigurationFuture).Future.OracleSpecBinding.TradingTerminationProperty = "vegaprotocol.builtin.timestamp"
 
 	// setup
 	eng.ensureTokenBalanceForParty(t, proposer, 1000)

